### ISSUE TITLE
replicate folders + albums from one library to another library (shared library)

### ISSUE THREAD

**Issue by mschindl** [RhetTbull/osxphotos#1683]
**Is your feature request related to a problem? Please describe.**
Is there a way replicate albums from one library to another library?
Shared library on both different devices and iCloud accounts in use and 95% of all pictures are stored there.
This feature request came from the use of a shared family library and currently only ONE iCloud has the folder + album structure.
The structure should be available on all devices.

**Describe the solution you'd like**
1. Using one shared library with a structure by folders and albums
2. Export information about folder + album with related picture link
3. Import information about folder + album with related picture link
4. Second library got the same structure

**Describe alternatives you've considered**
Nothing found yet

**Comment by oPromessa** [comment]
Hello @mschindl   

See if this ask is identical to yours #887

From it, @RhetTbull built the [osxphotos sync option](https://rhettbull.github.io/osxphotos/cli.html#osxphotos-sync)

- Only catch is that it does not (yet!) create folder structure.

**Comment by RhetTbull** [comment]
Hi @mschindl OSXPhotos has two ways to do what you're looking for. You can use the `osxphotos sync` or `osxphotos batch-edit` commands to sync albums between libraries, with some limitations described below.

The first is the `osxphotos sync` command as @oPromessa pointed out. This can sync albums but unfortunately does not yet sync folders (this will eventually be fixed, see #1639).  The sync command matches photos between the libraries via signature (what Photos internally calls a fingerprint). The photos must be exact duplicates to be matched between the two libraries. You can read more about the sync command by running `osxphotos help sync`.  Here's a snippet from the docs.  If you only want to sync albums and not other metadata, you'd use `--merge albums` in place of `--merge all`. This method requires you to run the sync command on both computers and to share the sync database or copy it between computers.

```
You can run the --export and --import commands together. In this case, the
  import will be run first and then the export will be run.

  For example, if you want to sync two Photos libraries between users or two
  different computers, you can export the metadata to a shared folder.

  On the first computer, run:

  osxphotos sync --export /path/to/export/folder/computer1.db --merge all
  --import /path/to/export/folder/computer2.db

  On the second computer, run:

  osxphotos sync --export /path/to/export/folder/computer2.db --merge all
  --import /path/to/export/folder/computer1.db
```

The second method uses the `osxphotos batch-edit` command to turn albums and folders into specially formatted keywords which are synced across iCloud shared library automatically by Photos. This method does successfully sync folders in addition to albums. Run `osxphotos help batch-edit` to learn more about the command.

If you run the following command periodically on each user's Mac, it will create keywords in the form of `album:Folder/Subfolder/Album` for each photo in an album. Then it will find any such keywords that already exist and add each photo to the appropriate album, creating folders and albums as necessary. You must select the photos to apply this to (e.g. all those in an album or all those in the shared iCloud library). 

osxphotos batch-edit --verbose --keyword "{album?album:{folder_album}}" --album "{keyword|filter(startswith album:)|sslice(6:)}" --split-folder "/"

This command uses the osxphotos template language to create keywords if an album exists and then uses the template language to extract the folder and album names from the keywords which are then used to create the albums (if needed) and then add the photos to the appropriate albums.

There is a limitation that if you remove a photo from an album, the keywords are not removed nor is the photo removed from the album. AppleScript, which OSXPhotos uses to add photos to the album, is not able to remove photos from albums.  Photos will all be tagged with these new keywords with no easy way to remove them so keep that in mind before you run this.

**Comment by mschindl** [comment]
Great explanation and work. It sounds like osxphotos batch-edit will solve my request.
I will test it asap and give feedback.

Many thanks

**Comment by RhetTbull** [comment]
Yes it will work for multiple albums. Each album would result in a new keyword.

**Comment by mschindl** [comment]
Great, it works good also with special characters in the names.
Is there a way to speed up batch-edit?

I'm using following command and removing "--verbose" will not speed up the process.
"osxphotos batch-edit --verbose --keyword "{album?album:{folder_album}}" --album "{keyword|filter(startswith album:)|sslice(6:)}" --split-folder "/" > ~/file-output.txt

In example it runs for over 10hours for round about 50k pictures/videos.

**Comment by RhetTbull** [comment]
This is because the AppleScript interface to Photos is really really slow and that's the only way to modify keywords. If Apple provided a better API, I'd use it but they seem uninterested in allowing users more control over the Photos library. 

One possible optimization would be to first read all the existing folders and keywords using OSXPhotos then only use AppleScript for the ones that need updating. First run would be slow but subsequent runs would be faster. 

Once you do the initial run, you can select only photos you've worked with recently or only albums you've added to then run the command on a smaller set of photos.

**Comment by mschindl** [comment]
I executed the command on an **other** library (without any folder/album) and if I select only a few objects (pic/vid) then it works well, but if I select the complete library with round about 47k object then I got the message "No photos selected".

Any idea why the error is shown? Is it a scaling issue?

**Comment by RhetTbull** [comment]
My guess is AppleScript is timing out getting the selected objects. You'll have to try with fewer photos to see what the limit is.

**Comment by RhetTbull** [comment]
It may take me a few days but I'll put together a script to remove the "album:" keywords in case you want to back out of this workflow. AppleScript limitations may prevent this from being tenable.

**Comment by mschindl** [comment]
If I understand it right then the usage of “album:” keywords will make trouble with future Photos release, right?
Tags can be easily removed with the UI/App Photos in a very fast way. 

What is your recommendation to use now?

I used batch-edit successfully now, but of course it tooks time.
50k objects in round about 20hours.

**Comment by mschindl** [comment]
I want to give some feedback: Unfortunately it does not scale, because after doing the batch job on the 2nd macOS for 30k objects it will not work for more as 250 objects now. (20k objects left)

**Comment by RhetTbull** [comment]
It will be some time before I can incorporate it but I've been doing some hacking on a side project that uses private APIs to do this type of thing natively. I did a test with ~1000 assets and results are much more promising:

```bash
$  python test.py
Added keyword to 1087 assets in 104.11 seconds
Added 1087 assets to album in 1.19 seconds
```

**Comment by mschindl** [comment]
I tried the "sync" command today because my folder structure is stable and does not change that often.

Unfortunately, it creates a new/same album at the top level ("My Albums") and does not synchronize the images in the current folder/albumXYZ where the album already exists or needs to be created. I thought there was only an issue with folder creation.

Btw: "-i" is not working and "--import" have to be used, because following error will occur
`Error: Got unexpected extra argument (XYZ.db)`
`osxphotos sync -e ~/ZZZ.db --merge all -i ~/YYY.db --dry-run`


I appreciate your work and I hope to help you with my tests and reports.

**Comment by RhetTbull** [comment]
> Btw: "-i" is not working and "--import" have to be used, because following error will occur

This is a bug caused because `-i` is redefined in the arguments:

`-i, --ignore-case  Case insensitive search for title, description, place, keyword, person, or album.`

I will open a new issue for this.

> Unfortunately, it creates a new/same album at the top level ("My Albums") and does not synchronize the images in the current folder/albumXYZ where the album already exists or needs to be created. I thought there was only an issue with folder creation.

I think this is the same as issue #1639. It's nearing the top of my queue but I've been blocked by limited time and working on same major changes to the date/time code in OSXPhotos that will improve handling of timezones but has requited extensive testing and updates to the test suite.  Currently, the sync code only considers album names, not folders.

**Comment by mschindl** [comment]
> Thanks for the additional data. This is helpful. It will be some time before I can incorporate it but I've been doing some hacking on a side project that uses private APIs to do this type of thing natively. I did a test with ~1000 assets and results are much more promising:
> 
> ```
> $  python test.py
> Added keyword to 1087 assets in 104.11 seconds
> Added 1087 assets to album in 1.19 seconds
> ```

Have you made progress? Can I give you more information to help you move forward?

**Comment by RhetTbull** [comment]
This has been on the back burner. I've been working on a rewrite of the date/time handling code to fix some bug and improve some features of the `timewarp` command.  That's almost done and now I've got to switch to python 3.13 support because osxphotos installation in homebrew is now broken as homebrew insists on using python 3.13 for pipx.

**Comment by RhetTbull** [comment]
>  I've been doing some hacking on a side project that uses private APIs to do this type of thing natively. I did a test with ~1000 assets and results are much more promising:

Looks like Apple may have broken 3rd party access to the private APIs in macOS 15.2

**Comment by mschindl** [comment]
Oh, bad news. I was hoping to work with your project in the future.
I hope you can find a workaround, because your project fills the gap with a feature that does not currently exist - shared library in a family.
🤞🏼

**Comment by RhetTbull** [comment]
If you know Python, it would be possible to write a script that modifies the keywords only for photos in a given album (but not a Smart Album until I implement that which requires a fair bit of work). The [cli_example_1.py](https://github.com/RhetTbull/osxphotos/blob/main/examples/cli_example_1.py) script provides a good starting point for a script that works with all the `osxphotos query` flags, for example: `osxphotos run myscript.py --album "Photos to update"`

**Comment by mschindl** [comment]
Some time flies away and @RhetTbull I have one idea if we have a scaling issue.

Could it help to limit the procession time frame per year of pictures/videos?
Or do we still have an issue with _"Apple may have broken 3rd party access to the private APIs in macOS 15.2"_

In example for import: **Process pictures from "--year 2024"**

  On the first computer, run:

  osxphotos sync --export /path/to/export/folder/computer1.db --merge all
  --import /path/to/export/folder/computer2.db

  On the second computer, run:

  osxphotos sync --export /path/to/export/folder/computer2.db --merge all
  --import /path/to/export/folder/computer1.db **--year 2024**

**Comment by RhetTbull** [comment]
> Or do we still have an issue with "Apple may have broken 3rd party access to the private APIs in macOS 15.2"

This was fixed in 15.3 so, for now, this is an option. It also appears to be still working in the Tahoe beta.

> In example for import: Process pictures from "--year 2024"

This is a good idea. Perhaps also `--from-date` and `--to-date` which are standard osxphotos query options.

**Comment by mschindl** [comment]
Looks good, but the folder issue (not created/recognised) is still available.
Can I use --from-date and --to-date with batch-edit in any way?

Keywords adding/checking/processing is slow, but if I'm not running in any timeouts while I defined only one year then it's good enough :)

**Comment by RhetTbull** [comment]
Unfortunately batch-edit won't let you fix albums/folders across different libraries and it doesn't currently include `--from-date`, `--to-date`. I am currently adding those options (along with `--uuid`, `--uuid-from-file`, and `--album`) to `batch-edit` to enable easier selection of photos for processing. (Should be out this week). In the mean time, you can create a smart album of the photos and then select them and run `batch-edit` (this works on macOS 15 but on earlier versions, selection in a smart album may not work)

**Comment by RhetTbull** [comment]
My current working branch adds the following options to batch-edit and I'll add them to timewarp before next release:

`--album, --folder, --from-date, --to-date, --year, --added-before, --added-after, --added-in-last, --uuid, --uuid-from-file`

This is a subset of the full query options that should make it easier to select the group of photos to operate on for commands that modify data in the library. If omitted, the commands use the current selection.

Once I merge that branch, I'll take a look at fixing the full folder/album path sync issue.

**Comment by oPromessa** [comment]
> Unfortunately, it creates a new/same album at the top level ("My Albums") and does not synchronize the images in the current folder/albumXYZ where the album already exists or needs to be created. I thought there was only an issue with folder creation.

@mschindl Don't know if this topic is still open for you. 

The last versions of osxphotos 0.74.2 now recreate the folder and album structure on `sync`. Give it a go and let us know.

**Comment by mschindl** [comment]
@oPromessa the topic is still open for me and now I had time to test it.

First test shows it takes time, because it syncs one picture per second and it seems so work now ;)
That's great. Thank you very much!

On the first computer - I executed following command
osxphotos sync --export /path/to/export/folder/computer1.db

On the second computer - I executed following command:
osxphotos sync --export /path/to/export/folder/computer2.db --merge all
--import /path/to/export/folder/computer1.db --year 2026

---
One question:

I did this for the entire library without "--year" and an existing album was created again > two time in "vacations" folder.
Could that be a bug?

**Comment by oPromessa** [comment]
Hello @mschindl. Sort for some delay. 

May you please elaborate?

When syncing in computer B it ends up creating the same album twice in the same folder:

Vacations
-- Some Album
---- images
-- Some Album
---- images


Are all images on both albums?
Does "Some Album" have international or special characters (like ã or $ or &)?

Can you share maybe the output of sync with --verbose option on?
-- do make sure not to share any private information.
